File: httpserver/src/main/java/com/shengshijie/httpserver/HttpServer.java
package com.shengshijie.httpserver;

import android.util.Log;

import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioChannelOption;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.handler.codec.http.HttpObjectAggregator;
import io.netty.handler.codec.http.HttpServerCodec;

public class HttpServer {

    public static void start(int port) {
        ServerBootstrap bootstrap = new ServerBootstrap();
        EventLoopGroup bossGroup = new NioEventLoopGroup();
        EventLoopGroup workerGroup = new NioEventLoopGroup();
        final HttpHandler httpHandler = new HttpHandler();
        httpHandler.registerRouter();
        bootstrap.group(bossGroup, workerGroup);
        bootstrap.channel(NioServerSocketChannel.class);
        bootstrap.childOption(NioChannelOption.TCP_NODELAY, true);
        bootstrap.childOption(NioChannelOption.SO_REUSEADDR, true);
        bootstrap.childOption(NioChannelOption.SO_KEEPALIVE, false);
        bootstrap.childOption(NioChannelOption.SO_RCVBUF, 2048);
        bootstrap.childOption(NioChannelOption.SO_SNDBUF, 2048);
        bootstrap.childHandler(new ChannelInitializer<SocketChannel>() {
            @Override
            public void initChannel(SocketChannel ch) {
                ch.pipeline().addLast("codec", new HttpServerCodec());
                ch.pipeline().addLast("aggregator", new HttpObjectAggregator(512 * 1024));
                ch.pipeline().addLast("logging", new FilterLoggingHandler());
                ch.pipeline().addLast("interceptor", new InterceptorHandler());
                ch.pipeline().addLast("bizHandler", httpHandler);
            }
        })
        ;
        ChannelFuture channelFuture = bootstrap.bind(port).syncUninterruptibly().addListener(future -> {
            Log.i("SERVER", "Server Start");
        });
        channelFuture.channel().closeFuture().addListener(future -> {
            Log.i("SERVER", "Server Shutdown");
            bossGroup.shutdownGracefully();
            workerGroup.shutdownGracefully();
        });
    }

}

File: httpserver/src/main/java/com/shengshijie/httpserver/HttpServer.kt
package com.shengshijie.httpserver

import com.shengshijie.httpserver.file.FileUploadHandler
import com.shengshijie.httpserver.http.FilterLoggingHandler
import com.shengshijie.httpserver.http.HttpHandler
import com.shengshijie.httpserver.http.InterceptorHandler
import com.shengshijie.httpserver.websocket.WebSocketHandler
import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.ChannelInitializer
import io.netty.channel.EventLoopGroup
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.SocketChannel
import io.netty.channel.socket.nio.NioChannelOption
import io.netty.channel.socket.nio.NioServerSocketChannel
import io.netty.handler.codec.http.HttpObjectAggregator
import io.netty.handler.codec.http.HttpRequestDecoder
import io.netty.handler.codec.http.HttpResponseEncoder
import io.netty.handler.logging.LogLevel
import io.netty.handler.stream.ChunkedWriteHandler


object HttpServer {

    private var bossGroup: EventLoopGroup? = null
    private var workerGroup: EventLoopGroup? = null
    private var running: Boolean = false

    @JvmStatic
    @JvmOverloads
    fun start(port: Int, log: (level: LogLevel, content: String) -> Unit = { _, _ -> }) {
        if (running) {
            log(LogLevel.INFO, "server is already running")
            return
        }
        try {
            val bootstrap = ServerBootstrap()
            bossGroup = NioEventLoopGroup()
            workerGroup = NioEventLoopGroup()
            bootstrap.group(bossGroup, workerGroup)
            bootstrap.channel(NioServerSocketChannel::class.java)
            bootstrap.childOption(NioChannelOption.TCP_NODELAY, true)
            bootstrap.childOption(NioChannelOption.SO_REUSEADDR, true)
            bootstrap.childOption(NioChannelOption.SO_KEEPALIVE, false)
            bootstrap.childOption(NioChannelOption.SO_RCVBUF, 2048)
            bootstrap.childOption(NioChannelOption.SO_SNDBUF, 2048)
            bootstrap.childHandler(object : ChannelInitializer<SocketChannel>() {
                public override fun initChannel(ch: SocketChannel) {
                    ch.pipeline().addLast("decoder", HttpRequestDecoder())
                    ch.pipeline().addLast("encoder", HttpResponseEncoder())
                    ch.pipeline().addLast("aggregator", HttpObjectAggregator(1024 * 1024 * 5))
                    ch.pipeline().addLast("chunked", ChunkedWriteHandler())
                    ch.pipeline().addLast("logging", FilterLoggingHandler(log))
                    ch.pipeline().addLast("interceptor", InterceptorHandler())
                    ch.pipeline().addLast("dispatchHandler", DispatchHandler())
                    ch.pipeline().addLast("httpHandler", HttpHandler().apply { registerRouter() })
                    ch.pipeline().addLast("webSocketHandler", WebSocketHandler())
                    ch.pipeline().addLast("fileUploadHandler", FileUploadHandler())
                }
            })
            val channelFuture = bootstrap.bind(port).syncUninterruptibly().addListener { log(LogLevel.INFO, "server start at port $port") }
            running = true
            channelFuture.channel().closeFuture().addListener {
                bossGroup?.shutdownGracefully()
                workerGroup?.shutdownGracefully()
            }
        } catch (e: Exception) {
            log(LogLevel.INFO, "server start error:" + e.message)
        }
    }

    @JvmStatic
    fun stop(log: (level: LogLevel, content: String) -> Unit = { _, _ -> }) {
        if (!running) {
            log(LogLevel.INFO, "server is already stop")
            return
        }
        try {
            bossGroup?.shutdownGracefully()
            workerGroup?.shutdownGracefully()
            bossGroup = null
            workerGroup = null
            running = false
            log(LogLevel.INFO, "server stop")
        } catch (e: Exception) {
            log(LogLevel.INFO, "server stop error:" + e.message)
        }
    }

}